cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AbstractMonitorHandler.java

```diff
@@ -207,31 +207,52 @@ private int getSystemServiceInstanceCount(String serviceName) throws Exception {
   }
 
   public void restartAllServiceInstances(HttpRequest request, HttpResponder responder, String serviceName) {
-    restartInstances(responder, serviceName, -1);
+    restartInstances(responder, serviceName, -1, true);
   }
 
   public void restartServiceInstance(HttpRequest request, HttpResponder responder, String serviceName,
                                      int instanceId) {
-    restartInstances(responder, serviceName, instanceId);
+    restartInstances(responder, serviceName, instanceId, false);
   }
 
-  private void restartInstances(HttpResponder responder, String serviceName, int instanceId) {
+  private void restartInstances(HttpResponder responder, String serviceName, int instanceId, boolean restartAll) {
     long startTimeMs = System.currentTimeMillis();
     boolean isSuccess = true;
+    if (!serviceManagementMap.containsKey(serviceName)) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
+      return;
+    }
+    MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
+
     try {
-      if (!serviceManagementMap.containsKey(serviceName)) {
-        responder.sendString(HttpResponseStatus.NOT_FOUND, String.format("Invalid service name %s", serviceName));
-        return;
+      if (!masterServiceManager.isServiceEnabled()) {
+        throw new IllegalStateException();
       }
 
-      MasterServiceManager masterServiceManager = serviceManagementMap.get(serviceName);
-      if (instanceId < 0) {
+      if (restartAll) {
         masterServiceManager.restartAllInstances();
       } else {
+        if (instanceId < 0 || instanceId >= masterServiceManager.getInstances()) {
+          throw new IllegalArgumentException();
+        }
         masterServiceManager.restartInstances(instanceId);
       }
-
       responder.sendStatus(HttpResponseStatus.OK);
+    } catch (IllegalStateException ise) {
+      LOG.warn(String.format("IllegalStateException when trying to restart instances for service %s", serviceName));
+
+      isSuccess = false;
+      responder.sendString(HttpResponseStatus.FORBIDDEN,
+                           String.format("Fail to restart instance for % because the service may not be enabled or " +
+                                           "not ready yet", serviceName));
+    } catch (IllegalArgumentException iex) {
+      LOG.warn(String.format("IllegalArgumentException when trying to restart instances for service %s", serviceName),
```

Comment:
Actually, you are right lets change it as debug since there was no logging before and not useful for admin.